File: java_bindings/lib/src/main/java/arpirobot/arduino/sensor/IRReflectorModule.java
/*
 * Copyright 2021 Marcus Behel
 *
 * This file is part of ArPiRobot-CoreLib.
 * 
 * ArPiRobot-CoreLib is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * ArPiRobot-CoreLib is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public License
 * along with ArPiRobot-CoreLib.  If not, see <https://www.gnu.org/licenses/>. 
 */

package arpirobot.arduino.sensor;

import arpirobot.Bridge;
import arpirobot.arduino.device.ArduinoDevice;

/**
 * Infrared reflection detector module
 */
public class IRReflectorModule extends ArduinoDevice {

    /**
     * @param digitalPin The digital output pin for this sensor
     */
    public IRReflectorModule(int digitalPin){
        this("" + digitalPin);
    }

    /**
     * @param digitalPin The digital output pin for this sensor (prefix with letter A for analog pin)
     */
    public IRReflectorModule(String digitalPin){
        this(digitalPin, "");
    }

    /**
     * @param digitalPin The digital output pin for this sensor
     * @param analogPin The analog output pin for this sensor (can be prefixed with letter A, but not required)
     */
    public IRReflectorModule(int digitalPin, String analogPin){
        this("" + digitalPin, "" + analogPin);
    }

    /**
     * @param digitalPin The digital output pin for this sensor (prefix with letter A for analog pin)
     * @param analogPin The analog output pin for this sensor (can be prefixed with letter A, but not required)
     */
    public IRReflectorModule(String digitalPin, String analogPin){
        this(digitalPin, analogPin, true, -1);
    }

    /**
     * @param digitalPin The digital output pin for this sensor
     * @param createDevice Leave this true unless the device is hard-coded in arduino firmware
     * @param deviceId Set this to the hard-coded deviceId if createDevice is false
     */
    public IRReflectorModule(int digitalPin, boolean createDevice, int deviceId){
        this("" + digitalPin, createDevice, deviceId);
    }

    /**
     * @param digitalPin The digital output pin for this sensor
     * @param analogPin The analog output pin for this sensor (can be prefixed with letter A, but not required)
     * @param createDevice Leave this true unless the device is hard-coded in arduino firmware
     * @param deviceId Set this to the hard-coded deviceId if createDevice is false
     */
    public IRReflectorModule(int digitalPin, String analogPin, boolean createDevice, int deviceId){
        this("" + digitalPin, analogPin, createDevice, deviceId);
    }

    /**
     * @param digitalPin The digital output pin for this sensor (prefix with letter A for analog pin)
     * @param createDevice Leave this true unless the device is hard-coded in arduino firmware
     * @param deviceId Set this to the hard-coded deviceId if createDevice is false
     */
    public IRReflectorModule(String digitalPin, boolean createDevice, int deviceId){
        this(digitalPin, "", createDevice, deviceId);
    }

    /**
     * @param digitalPin The digital output pin for this sensor (prefix with letter A for analog pin)
     * @param analogPin The analog output pin for this sensor (can be prefixed with letter A, but not required)
     * @param createDevice Leave this true unless the device is hard-coded in arduino firmware
     * @param deviceId Set this to the hard-coded deviceId if createDevice is false
     */
    public IRReflectorModule(String digitalPin, String analogPin, boolean createDevice, int deviceId){
        ptr = Bridge.arpirobot.IRReflectorModule_create(digitalPin, analogPin, createDevice, deviceId);
    }

    @Override
    protected void doDestroy() {
        Bridge.arpirobot.IRReflectorModule_destroy(ptr);
    }

    /**
     * Get the digital value read from the sensor
     * @return true or false. One cooresponds to reflection detected. The other is no reflection detected.
     */
    public boolean getDigitalValue(){
        verifyNotDestroyed();
        return Bridge.arpirobot.IRReflectorModule_getDigitalValue(ptr);
    }

    /**
     * Get the analog reflection signal strength (if this device was setup to use an analog input)
     * @return The read analog value (will be zero if not configured for analog data)
     */
    public int getAnalogValue(){
        verifyNotDestroyed();
        return Bridge.arpirobot.IRReflectorModule_getAnalogValue(ptr);
    }
}

File: cpp_library/src/arpirobot/arduino/sensor/IRReflectorModule.cpp
/*
 * Copyright 2021 Marcus Behel
 *
 * This file is part of ArPiRobot-CoreLib.
 * 
 * ArPiRobot-CoreLib is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * ArPiRobot-CoreLib is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public License
 * along with ArPiRobot-CoreLib.  If not, see <https://www.gnu.org/licenses/>. 
 */

#include <arpirobot/arduino/sensor/IRReflectorModule.hpp>
#include <arpirobot/core/log/Logger.hpp>
#include <arpirobot/core/conversions.hpp>

#include <stdexcept>

using namespace arpirobot;


IRReflectorModule::IRReflectorModule(int digitalPin, bool createDevice, int deviceId) : 
        ArduinoDevice(createDevice, deviceId), digitalPin(std::to_string(digitalPin)), analogPin(""){
    if(analogPin.length() == 0){
        deviceName = "IRReflectorModule(" + this->digitalPin + ")";
    }else{
        deviceName = "IRReflectorModule(" + this->digitalPin + ", " + this->analogPin + ")";
    }
}

IRReflectorModule::IRReflectorModule(std::string digitalPin, bool createDevice, int deviceId) : 
        ArduinoDevice(createDevice, deviceId), digitalPin(digitalPin), analogPin(""){
    if(analogPin.length() == 0){
        deviceName = "IRReflectorModule(" + this->digitalPin + ")";
    }else{
        deviceName = "IRReflectorModule(" + this->digitalPin + ", " + this->analogPin + ")";
    }
}

IRReflectorModule::IRReflectorModule(int digitalPin, std::string analogPin, bool createDevice, int deviceId) : 
        ArduinoDevice(createDevice, deviceId), digitalPin(std::to_string(digitalPin)), analogPin(analogPin){
    if(analogPin.length() == 0){
        deviceName = "IRReflectorModule(" + this->digitalPin + ")";
    }else{
        deviceName = "IRReflectorModule(" + this->digitalPin + ", " + this->analogPin + ")";
    }     
}

IRReflectorModule::IRReflectorModule(std::string digitalPin, std::string analogPin, bool createDevice, int deviceId)  : 
        ArduinoDevice(createDevice, deviceId), digitalPin(digitalPin), analogPin(analogPin){
    if(analogPin.length() == 0){
        deviceName = "IRReflectorModule(" + this->digitalPin + ")";
    }else{
        deviceName = "IRReflectorModule(" + this->digitalPin + ", " + this->analogPin + ")";
    }
}

bool IRReflectorModule::getDigitalValue(){
    return digitalValue;
}

int IRReflectorModule::getAnalogValue(){
    return analogValue;
}

void IRReflectorModule::applyDefaultState(){
    digitalValue = false;
    analogValue = 0;
}

std::vector<uint8_t> IRReflectorModule::getCreateData(){
    bool isDigPinAnalog = digitalPin[0] == 'A' || digitalPin[0] == 'a';
    int dp = 0, ap = 0;
    try{
        if(isDigPinAnalog){
            dp = std::stoi(digitalPin.substr(1, digitalPin.length() - 1));
        }else{  
            dp = std::stoi(digitalPin);
        }
    }catch(const std::runtime_error &e){
        Logger::logWarningFrom(getDeviceName(), "Invalid digital pin number");
        return {};
    }
    if(analogPin.length() == 0){
        ap = 255; // 255 means analog disabled
    }else{
        try{
            // Allow analog pin to either start with 'A' or not
            if(analogPin[0] == 'A' || analogPin[0] == 'a'){
                ap = std::stoi(analogPin.substr(1, analogPin.length() - 1));
            }else{
                ap = std::stoi(analogPin);
            }
        }catch(const std::runtime_error &e){
            Logger::logWarningFrom(getDeviceName(), "Invalid analog pin number");
        }
    }

    // ADDIRREFLECTOR,is_dig_analog,digital_pin,analog_pin
    std::vector<uint8_t> data = stringToData("ADDIRREFLECTOR");
    data.push_back(isDigPinAnalog);
    data.push_back(dp);
    data.push_back(ap);
    return data;
}

void IRReflectorModule::handleData(const std::vector<uint8_t> &data){
    // data = deviceId, data..., crc, crc
    if(data.size() >= 6){
        // Has digital and analog values
        digitalValue = data[1];
        analogValue = Conversions::convertDataToInt16(data, 2, true);
    }else if(data.size() >= 4){
        // Has only digital value
        digitalValue = data[1];
    }
}
